### PR TITLE
fix(migration): move description model validation after domain

### PR DESCRIPTION
**Pull request by nvinuesa**
Currently, after filling the `description.Model` when migrating the legacy db contents we validate it. This happens before we start filling it with the contents from dqlite.

We face an issue here, that on some cases the validation will fail because some of the missing data will arrive after we finish the dqlite stage of the migration. This is the case for machines/instance data migration.

We therefore move the whole Validation() to internal/migration, *after* dqlite.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [ ] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

Migration should not be affected:

```
juju bootstrap localhost tgt
juju bootstrap localhost src
juju add-model m
juju deploy ubuntu
# Wait until ACTIVE
juju migrate m tgt
```
Also, https://github.com/juju/juju/pull/18214 should pass the migration checks after rebased on top of this one.

## Links

**Jira card:** JUJU-

### CODE REVIEW

**Review by jack-w-shaw — APPROVED** [review]
LGTM

File: internal/migration/migration.go [hunk 1]
```diff
@@ -113,6 +113,11 @@ func (e *ModelExporter) Export(ctx context.Context, model description.Model) (de
 	if err := coordinator.Perform(ctx, e.scope, model); err != nil {
 		return nil, errors.Trace(err)
 	}
+	// The model now contains all the exported data from the legacy state along
+	// with the new domains' one. Time to validate.
+	if err := model.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
```
Comment (SimonRichardson) [inline internal/migration/migration.go:120]: We should add a test for this, to ensure it's triggered.

File: state/migration_export.go [hunk 2]
```diff
@@ -204,14 +204,6 @@ func (st *State) exportImpl(cfg ExportConfig, leaders map[string]string, store o
...
 		return nil, errors.Trace(err)
 	}
-	// If we are doing a partial export, it doesn't really make sense
-	// to validate the model.
-	fullExport := ExportConfig{}
-	if cfg == fullExport {
-		if err := export.model.Validate(); err != nil {
-			return nil, errors.Trace(err)
-		}
-	}
```
Comment (SimonRichardson) [inline state/migration_export.go:214]: I'm saddened that we could do this without any test failures.

File: internal/migration/migration.go (hunk 1 above)
Comment (nvinuesa) [reply internal/migration/migration.go]: In the last commit I did a bit of refactoring so we can inject dependencies on the internal migration package, and therefore be able to correctly test this method (at least the Validation stage).

So the test is now there.

**Review by SimonRichardson — COMMENTED** [review]
Just a slight tweak. Normally we don't really care the order of mocks, but we *do* care here. The `validate` call must be the last call in the export request.

File: internal/migration/migration_test.go [hunk 3]
```diff
@@ -37,6 +37,73 @@ import (
...
+		nil,
+		scope,
+		s.storageRegistryGetter,
+		s.coordinator,
+		nil, nil,
+	)
+
+	s.operationsExporter.EXPECT().ExportOperations(s.providerRegistry)
+	s.model.EXPECT().Validate().Return(nil)
+	s.coordinator.EXPECT().Perform(gomock.Any(), scope, s.model).Return(nil)
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply internal/migration/migration_test.go]: Nice catch! Fixed it.

File: domain/modelmigration/export.go [hunk 4]
```diff
@@ -23,26 +23,42 @@ import (
 	internalstorage "github.com/juju/juju/internal/storage"
 )
 
-// ExportOperations registers the export operations with the given coordinator.
-// This is a convenience function that can be used by the main migration package
-// to register all the export operations.
-func ExportOperations(
+type Exporter struct {
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply domain/modelmigration/export.go:29]: Comment added, nice catch!

**Review by SimonRichardson — APPROVED** [review]
Thanks.

**Pull request merged**